server/src/main/java/org/elasticsearch/repositories/Repository.java

```diff
@@ -225,17 +226,18 @@ void restoreShard(Store store, SnapshotId snapshotId, IndexId indexId, ShardId s
 
     /**
      * Execute a cluster state update with a consistent view of the current {@link RepositoryData}. The {@link ClusterState} passed to the
-     * task generated through {@code createUpdateTask} is guaranteed to point at the same state for this repository as the did the state
-     * at the time the {@code RepositoryData} was loaded.
+     * task generated through {@code createUpdateTaskAsync} is guaranteed to point at the same state for this repository as the
+     * state at the time the {@code RepositoryData} was loaded.
      * This allows for operations on the repository that need a consistent view of both the cluster state and the repository contents at
      * one point in time like for example, checking if a snapshot is in the repository before adding the delete operation for it to the
      * cluster state.
      *
-     * @param createUpdateTask function to supply cluster state update task
-     * @param source           the source of the cluster state update task
-     * @param onFailure        error handler invoked on failure to get a consistent view of the current {@link RepositoryData}
+     * @param createUpdateTaskAsync async supplier of the cluster state update task
+     * @param source                the source of the cluster state update task
+     * @param onFailure             error handler invoked on failure to get a consistent view of the current {@link RepositoryData}
      */
-    void executeConsistentStateUpdate(Function<RepositoryData, ClusterStateUpdateTask> createUpdateTask, String source,
+    void executeConsistentStateUpdate(BiConsumer<RepositoryData, ActionListener<ClusterStateUpdateTask>> createUpdateTaskAsync,
+                                      String source,
```

Comment:
The encrypted repository has to fork to produce the update task.
It does some expensive hashing that ought stay away from network threads.